File: pychivalry/log/diagnostics.py
"""
CK3 Log Diagnostics Converter - Convert log analysis to LSP diagnostics

DIAGNOSTIC CODES:
    LOGDIAG-001: URI resolution failed
    LOGDIAG-002: Diagnostic conversion error
    LOGDIAG-003: Publishing failed

MODULE OVERVIEW:
    This module converts log analysis results into LSP diagnostics that appear
    as red squiggles in the editor. It handles file path resolution, diagnostic
    merging with static analysis, and manages the diagnostic lifecycle.

ARCHITECTURE:
    **Conversion Flow**:
    ```
    LogAnalysisResult → Resolve file path → Create Diagnostic →
    Merge with existing → Publish to client
    ```
    
    **Diagnostic Sources**:
    - Static analysis (from pychivalry analyzers) - PRIMARY
    - Game logs (from this module) - SECONDARY
    
    Static analysis diagnostics take precedence and are not overwritten.

CLASSES:
    - LogDiagnosticConverter: Main converter and publisher

USAGE:
    ```python
    converter = LogDiagnosticConverter(server, workspace_root)
    
    # Convert and publish result
    diagnostic = converter.convert_to_diagnostic(result)
    if diagnostic:
        converter.publish_diagnostics(uri, [diagnostic])
    
    # Clear game log diagnostics
    converter.clear_log_diagnostics(uri)
    ```

FEATURES:
    - Path resolution (relative to workspace)
    - Diagnostic severity mapping
    - Source attribution ("ck3-game-log")
    - Merging with static analysis
    - Per-file diagnostic tracking
    - Bulk clear operations

PERFORMANCE:
    - Conversion: <1ms per diagnostic
    - Publishing: Batched for efficiency
    - Memory: ~100 bytes per diagnostic

DEPENDENCIES:
    - pygls>=2.0.0: LSP types and server
    - pathlib: Path manipulation (stdlib)

TESTING:
    See tests/test_log_diagnostics.py

AUTHOR:
    Cyborgninja21

VERSION:
    1.0.0 (2026-01-01)
"""

import logging
from pathlib import Path
from typing import TYPE_CHECKING, Dict, List, Optional
from urllib.parse import quote
from urllib.request import pathname2url

from lsprotocol import types

if TYPE_CHECKING:
    from pygls.server import LanguageServer
    from pychivalry.log_analyzer import LogAnalysisResult

logger = logging.getLogger(__name__)


class LogDiagnosticConverter:
    """
    Convert log analysis results to LSP diagnostics.
    
    This class manages the conversion of game log errors into LSP diagnostics
    that appear in the editor's Problems panel and as inline squiggles.
    
    **Features**:
    - Path resolution from workspace-relative paths
    - Diagnostic merging (doesn't override static analysis)
    - Source tracking (distinguishes game logs from static analysis)
    - Bulk operations (clear all log diagnostics)
    
    **Diagnostic Lifecycle**:
    1. Log error detected
    2. Convert to diagnostic
    3. Resolve file URI
    4. Merge with existing diagnostics
    5. Publish to client
    6. Track for later clearing
    
    Attributes:
        server: LSP server instance
        workspace_root: Root path of workspace
        log_diagnostics: Map of URI to log-sourced diagnostics
        
    Example:
        ```python
        converter = LogDiagnosticConverter(server, "/workspace/path")
        
        # Convert a result
        result = analyzer.analyze_line(line, "game.log")
        diagnostic = converter.convert_to_diagnostic(result)
        
        # Publish it
        if diagnostic and result.source_file:
            uri = converter._resolve_file_uri(result.source_file)
            if uri:
                converter.publish_diagnostics(uri, [diagnostic])
        ```
    """
    
    # Source identifier for game log diagnostics
    LOG_DIAGNOSTIC_SOURCE = "ck3-game-log"
    
    def __init__(self, server: "LanguageServer", workspace_root: str) -> None:
        """
        Initialize the diagnostic converter.
        
        Args:
            server: LSP server instance for publishing diagnostics
            workspace_root: Root directory of workspace for path resolution
            
        Example:
            ```python
            converter = LogDiagnosticConverter(
                server=my_server,
                workspace_root="/home/user/my_mod"
            )
            ```
        """
        self.server = server
        self.workspace_root = Path(workspace_root)
        
        # Track diagnostics we've published from logs (by URI)
        self.log_diagnostics: Dict[str, List[types.Diagnostic]] = {}
        
        logger.info(f"LogDiagnosticConverter initialized for workspace: {workspace_root}")
    
    def convert_to_diagnostic(self, result: "LogAnalysisResult") -> Optional[types.Diagnostic]:
        """
        Convert a log analysis result to an LSP diagnostic.
        
        Creates a diagnostic that will appear in the editor with appropriate
        severity, message, and source attribution.
        
        Args:
            result: LogAnalysisResult from analyzer
            
        Returns:
            types.Diagnostic if conversion successful, None otherwise
            
        Example:
            ```python
            result = analyzer.analyze_line(
                "Unknown effect: add_glod",
                "game.log"
            )
            diagnostic = converter.convert_to_diagnostic(result)
            if diagnostic:
                print(f"Severity: {diagnostic.severity}")
                print(f"Message: {diagnostic.message}")
            ```
            
        Notes:
            - Returns None if source file/line not extractable
            - Diagnostic range covers entire line
            - Source is always "ck3-game-log"
            - Code is category-based (e.g., "GAME_LOG_UNKNOWN_EFFECT")
        """
        # Must have source location
        if not result.source_file or result.line_number is None:
            logger.debug(f"Cannot create diagnostic: no source location in {result.category}")
            return None
        
        # Calculate range
        line = result.line_number - 1  # LSP is 0-indexed
        start_char = result.column_number or 0
        end_char = start_char + 100  # Cover whole line (will be clamped by client)
        
        # Create diagnostic
        diagnostic = types.Diagnostic(
            range=types.Range(
                start=types.Position(line=line, character=start_char),
                end=types.Position(line=line, character=end_char)
            ),
            message=f"[Game Log] {result.message}",
            severity=result.severity,
            source=self.LOG_DIAGNOSTIC_SOURCE,
            code=self._generate_diagnostic_code(result.category)
        )
        
        # Add suggestions as related information if available
        if result.suggestions:
            diagnostic.data = {
                "suggestions": result.suggestions,
                "category": result.category,
                "action_type": result.code_action_type
            }
        
        logger.debug(f"Created diagnostic: {diagnostic.code} at {result.source_file}:{result.line_number}")
        return diagnostic
    
    def publish_diagnostics(
        self,
        uri: str,
        new_diagnostics: List[types.Diagnostic],
        merge: bool = True
    ) -> None:
        """
        Publish diagnostics for a file.
        
        Optionally merges with existing diagnostics to avoid overwriting
        static analysis results.
        
        Args:
            uri: File URI to publish diagnostics for
            new_diagnostics: New diagnostics to add (from game logs)
            merge: Whether to merge with existing diagnostics (default: True)
            
        Example:
            ```python
            # Publish with merging (recommended)
            converter.publish_diagnostics(uri, [diagnostic])
            
            # Replace all diagnostics
            converter.publish_diagnostics(uri, [diagnostic], merge=False)
            ```
            
        Notes:
            - If merge=True, preserves existing diagnostics from other sources
            - Tracks log diagnostics for later clearing
            - Automatically called by LSP server methods
        """
        try:
            if merge:
                # Get existing diagnostics from server
                existing = self._get_existing_diagnostics(uri)
                
                # Filter out old log diagnostics
                non_log_diagnostics = [
                    d for d in existing
                    if d.source != self.LOG_DIAGNOSTIC_SOURCE
                ]
                
                # Combine with new log diagnostics
                all_diagnostics = non_log_diagnostics + new_diagnostics
            else:
                all_diagnostics = new_diagnostics
            
            # Track log diagnostics
            self.log_diagnostics[uri] = new_diagnostics.copy()
            
            # Publish to client
            self.server.publish_diagnostics(uri, all_diagnostics)
            
            logger.debug(f"Published {len(new_diagnostics)} log diagnostics for {uri}")
            
        except Exception as e:
            logger.error(f"Error publishing diagnostics for {uri}: {e}", exc_info=True)
    
    def clear_log_diagnostics(self, uri: str) -> None:
        """
        Clear all game log diagnostics for a file.
        
        Removes diagnostics from game logs while preserving static analysis
        diagnostics.
        
        Args:
            uri: File URI to clear log diagnostics from
            
        Example:
            ```python
            # Clear log diagnostics but keep static analysis
            converter.clear_log_diagnostics(uri)
            ```
            
        Notes:
            - Only removes LOG_DIAGNOSTIC_SOURCE diagnostics
            - Preserves all other diagnostic sources
        """
        try:
            # Get existing diagnostics
            existing = self._get_existing_diagnostics(uri)
            
            # Keep only non-log diagnostics
            non_log = [
                d for d in existing
                if d.source != self.LOG_DIAGNOSTIC_SOURCE
            ]
            
            # Publish updated list
            self.server.publish_diagnostics(uri, non_log)
            
            # Remove from tracking
            if uri in self.log_diagnostics:
                del self.log_diagnostics[uri]
            
            logger.debug(f"Cleared log diagnostics for {uri}")
            
        except Exception as e:
            logger.error(f"Error clearing log diagnostics for {uri}: {e}", exc_info=True)
    
    def clear_all_log_diagnostics(self) -> None:
        """
        Clear all game log diagnostics from all files.
        
        Useful when stopping log watching or resetting the session.
        
        Example:
            ```python
            # Stop watching and clear all log diagnostics
            watcher.stop()
            converter.clear_all_log_diagnostics()
            ```
        """
        logger.info(f"Clearing log diagnostics from {len(self.log_diagnostics)} files")
        
        for uri in list(self.log_diagnostics.keys()):
            self.clear_log_diagnostics(uri)
    
    def get_active_diagnostics(self) -> Dict[str, List[types.Diagnostic]]:
        """
        Get currently tracked log diagnostics.
        
        Returns:
            Dictionary mapping URIs to lists of log diagnostics
            
        Example:
            ```python
            active = converter.get_active_diagnostics()
            print(f"Log diagnostics in {len(active)} files")
            for uri, diags in active.items():
                print(f"  {uri}: {len(diags)} diagnostics")
            ```
        """
        return self.log_diagnostics.copy()
    
    def resolve_file_uri(self, file_path: str) -> Optional[str]:
        """
        Resolve a file path to a URI.
        
        Handles both absolute and workspace-relative paths.
        
        Args:
            file_path: File path (absolute or relative to workspace)
            
        Returns:
            File URI string or None if path invalid
            
        Example:
            ```python
            # Relative path
            uri = converter.resolve_file_uri("events/my_event.txt")
            # Returns: file:///workspace/events/my_event.txt
            
            # Absolute path
            uri = converter.resolve_file_uri("/absolute/path/file.txt")
            # Returns: file:///absolute/path/file.txt
            ```
        """
        try:
            path = Path(file_path)
            
            # If relative, make it relative to workspace root
            if not path.is_absolute():
                path = self.workspace_root / path
            
            # Check if path exists
            if not path.exists():
                logger.warning(f"File does not exist: {path}")
                # Still return URI - file might be created later
            
            # Convert to URI
            uri = path.as_uri()
            return uri
            
        except Exception as e:
            logger.error(f"Error resolving file URI for {file_path}: {e}", exc_info=True)
            return None
    
    def _get_existing_diagnostics(self, uri: str) -> List[types.Diagnostic]:
        """
        Get existing diagnostics for a file.
        
        Args:
            uri: File URI
            
        Returns:
            List of existing diagnostics
            
        Notes:
            - Tries to get from server workspace
            - Returns empty list if not available
        """
        try:
            # Try to get from workspace
            doc = self.server.workspace.get_text_document(uri)
            # Note: pygls doesn't store diagnostics on documents
            # We only have what we've published ourselves
            return self.log_diagnostics.get(uri, [])
        except Exception:
            return []
    
    def _generate_diagnostic_code(self, category: str) -> str:
        """
        Generate diagnostic code from category.
        
        Args:
            category: Error category string
            
        Returns:
            Diagnostic code string
            
        Example:
            _generate_diagnostic_code("unknown_effect")
            # Returns: "GAME_LOG_UNKNOWN_EFFECT"
        """
        return f"GAME_LOG_{category.upper()}"

File: vscode-extension/src/server/log/diagnostics.ts
/**
 * Log Diagnostics Converter — turns LogAnalysisResults into LSP Diagnostics
 *
 * Resolves CK3 log-reported file paths against workspace roots, creates
 * Diagnostic objects with `source: "ck3-game-log"`, and publishes them
 * without overwriting the static-analysis diagnostics already on file.
 *
 * DIAGNOSTIC CODES:
 *   GAME_LOG_{CATEGORY}  (e.g. GAME_LOG_UNKNOWN_EFFECT)
 */

import { Diagnostic, DiagnosticSeverity, Connection } from 'vscode-languageserver/node';
import * as path from 'path';
import * as fs from 'fs';
import { LogAnalysisResult } from './analyzer';
import { serverLogger } from '../utils/logger';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const LOG_DIAGNOSTIC_SOURCE = 'ck3-game-log';

// ---------------------------------------------------------------------------
// LogDiagnosticConverter
// ---------------------------------------------------------------------------

export class LogDiagnosticConverter {
    private connection: Connection;
    private workspaceRoots: string[];

    /** Log-sourced diagnostics keyed by document URI */
    private logDiagnostics = new Map<string, Diagnostic[]>();

    constructor(connection: Connection, workspaceRoots: string[]) {
        this.connection = connection;
        this.workspaceRoots = workspaceRoots;
    }

    /** Update the set of workspace roots (e.g. when folders change). */
    setWorkspaceRoots(roots: string[]): void {
        this.workspaceRoots = roots;
    }

    // -----------------------------------------------------------------------
    // Conversion
    // -----------------------------------------------------------------------

    /**
     * Convert a single LogAnalysisResult to an LSP Diagnostic.
     * Returns null if the result lacks a resolvable source location.
     */
    convertToDiagnostic(result: LogAnalysisResult): Diagnostic | null {
        if (!result.sourceFile || result.lineNumber === undefined) {
            return null;
        }

        // LSP lines are 0-indexed; CK3 log lines are 1-indexed
        const line = Math.max(0, result.lineNumber - 1);
        const character = result.columnNumber ? Math.max(0, result.columnNumber - 1) : 0;

        const diag: Diagnostic = {
            severity: result.severity,
            range: {
                start: { line, character },
                end: { line, character: character + 1 },
            },
            message: `[Game Log] ${result.message}`,
            source: LOG_DIAGNOSTIC_SOURCE,
            code: `GAME_LOG_${result.category.toUpperCase()}`,
        };

        // Attach suggestions as related data
        if (result.suggestions.length > 0) {
            (diag as any).data = {
                suggestions: result.suggestions,
                category: result.category,
                actionType: result.codeActionType,
            };
        }

        return diag;
    }

    /**
     * Convert a batch of results, group by resolved URI, and publish.
     * Merges with (does not overwrite) existing static-analysis diagnostics.
     */
    convertAndPublish(results: LogAnalysisResult[]): void {
        // Group diagnostics by URI
        const byUri = new Map<string, Diagnostic[]>();

        for (const result of results) {
            const uri = this.resolveFileUri(result.sourceFile);
            if (!uri) continue;

            const diag = this.convertToDiagnostic(result);
            if (!diag) continue;

            let list = byUri.get(uri);
            if (!list) {
                list = [];
                byUri.set(uri, list);
            }
            list.push(diag);
        }

        // Publish per-URI
        for (const [uri, diags] of byUri) {
            this.publishDiagnostics(uri, diags);
        }
    }

    // -----------------------------------------------------------------------
    // Publishing
    // -----------------------------------------------------------------------

    /**
     * Publish log diagnostics for a URI, merging with any already-tracked
     * log diagnostics (additive within a session).
     */
    publishDiagnostics(uri: string, newDiags: Diagnostic[]): void {
        const existing = this.logDiagnostics.get(uri) ?? [];
        const merged = [...existing, ...newDiags];
        this.logDiagnostics.set(uri, merged);

        // Send to client — only the log-sourced set (the server sends
        // static diagnostics separately and the client merges by source).
        this.connection.sendDiagnostics({ uri, diagnostics: merged });
    }

    /** Clear log diagnostics for a single URI. */
    clearLogDiagnostics(uri: string): void {
        this.logDiagnostics.delete(uri);
        this.connection.sendDiagnostics({ uri, diagnostics: [] });
    }

    /** Clear all tracked log diagnostics across all files. */
    clearAllLogDiagnostics(): void {
        for (const uri of this.logDiagnostics.keys()) {
            this.connection.sendDiagnostics({ uri, diagnostics: [] });
        }
        this.logDiagnostics.clear();
    }

    /** Return a read-only view of active log diagnostics. */
    getActiveDiagnostics(): ReadonlyMap<string, Diagnostic[]> {
        return this.logDiagnostics;
    }

    // -----------------------------------------------------------------------
    // Path resolution
    // -----------------------------------------------------------------------

    /**
     * Resolve a file path from a CK3 log entry to a `file://` URI.
     *
     * Tries the path as-is first (absolute), then relative to each
     * workspace root.  Returns null if no valid file is found.
     */
    resolveFileUri(filePath: string | undefined): string | null {
        if (!filePath) return null;

        // Normalise separators
        const normalized = filePath.replace(/\\/g, '/');

        // 1. Absolute path
        if (path.isAbsolute(normalized)) {
            if (this.fileExists(normalized)) {
                return this.pathToUri(normalized);
            }
        }

        // 2. Relative to workspace roots
        for (const root of this.workspaceRoots) {
            const candidate = path.resolve(root, normalized);
            if (this.fileExists(candidate)) {
                return this.pathToUri(candidate);
            }
        }

        return null;
    }

    private fileExists(p: string): boolean {
        try {
            fs.accessSync(p, fs.constants.R_OK);
            return true;
        } catch {
            return false;
        }
    }

    private pathToUri(p: string): string {
        const normalized = p.replace(/\\/g, '/');
        if (normalized.startsWith('/')) {
            return `file://${normalized}`;
        }
        return `file:///${normalized}`;
    }
}
